File: go/queue/255_my_stack.go
package queue

// 注释可以参考 stack 中的255题

type MyStack struct {
	queue1 []int
	queue2 []int
}

func Constructor() MyStack {
	return MyStack{}
}

func (ms *MyStack) Push(x int) {
	// queue 先入先出，要想实现stack 后入先出
	// 就需要把新进入的元素（即后入的元素）添加到queue 的头部
	// 加入queue2 的首部
	ms.queue2 = append(ms.queue2, x)
	// queue1 的元素从头到尾出队列
	for _, v := range ms.queue1 {
		ms.queue2 = append(ms.queue2, v)
		ms.queue1 = ms.queue1[1:]
	}
	// ms.queue2 = append(ms.queue2, ms.queue1...)

	// 交换queue1 和 queue2，保持永远从queue取数据
	ms.queue1, ms.queue2 = ms.queue2, ms.queue1
}

func (ms *MyStack) Pop() int {
	v := ms.queue1[0]
	ms.queue1 = ms.queue1[1:]
	return v
}

func (ms *MyStack) Top() int {
	return ms.queue1[0]
}

func (ms *MyStack) Empty() bool {
	return len(ms.queue1) == 0
}

File: rust/src/queue/255_my_stack.rs

// 请你仅使用两个队列实现一个后入先出（LIFO）的栈，并支持普通栈的全部四种操作（push、top、pop 和 empty）。

struct MyStack {
    queue1: Vec<i32>,
    queue2: Vec<i32>
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyStack {

    fn new() -> Self {
        MyStack { queue1: vec![], queue2: vec![] }
    }
    
    fn push(&mut self, x: i32) {
        self.queue2.push(x);

        while let Some(v) = self.queue1.first() {
            self.queue2.push(*v);
            self.queue1.remove(0);
        }

        std::mem::swap(&mut self.queue1, &mut self.queue2);
    }
    
    fn pop(&mut self) -> i32 {
        self.queue1.remove(0)
    }
    
    fn top(&self) -> i32 {
        *self.queue1.first().unwrap()
    }
    
    fn empty(&self) -> bool {
        self.queue1.is_empty()
    }
}